### PR TITLE
Add scalar subregisters for vector registers

### PR DESCRIPTION
**Pull request by nshp**
Rather than generating a bunch of shifts, truncates and ors for vector operations, this change creates and takes advantage of Binja's subregisters. This will make it much easier to lift actual vector operations (as opposed to scalar operations within vectors) in the future.

### CODE REVIEW

File: aarch64_extension.cpp [hunk 1]
```diff
@@ -97,6 +169,20 @@ class AArch64ArchitectureExtension : public ArchitectureHook {
     return (BNLowLevelILFlagCondition) -1;
   }
 
+  uint32_t getVectorSubregister(uint32_t baseRegister, size_t elemSize,
+                                size_t index) {
+    for (auto& regPair : mRegisters) {
```
Comment (meme) [inline aarch64_extension.cpp:174]: Is this going to be a chokepoint being O(n)?

Comment (meme) [inline aarch64_extension.cpp:133]: Is this going to be a chokepoint being O(n)?

File: aarch64_extension.cpp [hunk 2]
```diff
@@ -492,7 +623,7 @@ class AArch64ArchitectureExtension : public ArchitectureHook {
   bool GetInstructionLowLevelIL(const uint8_t* data, uint64_t addr, size_t& len,
                                 LowLevelILFunction& il) override {
     cs_insn* instr;
-    size_t count = cs_disasm(disassembler.Get(), data, len, addr, 0, &instr);
+    size_t count = cs_disasm(disassembler.Get(), data, 4, addr, 0, &instr);
```
Comment (meme) [inline aarch64_extension.cpp]: Why the change to `4`?

File: aarch64_extension.cpp [hunk 3]
```diff
@@ -369,20 +491,29 @@ class AArch64ArchitectureExtension : public ArchitectureHook {
...
                  -1 == detail->operands[1].vector_index && 16 == Rd_size &&
                  8 == Rn_size) {
         /* FMOV <Vd>.D[1], <Xn> */
-        il.AddInstruction(il.SetRegister(
-            Rd_size, Rd,
-            il.ZeroExtend(Rd_size, il.Or(Rn_size, il.Register(Rn_size, Rn),
-                                         il.Register(Rn_size, Rd)))));
+        uint32_t subRd = getVectorSubregister(Rd, Rn_size, 1);
+        if (0 == subRd) {
+          LogError("%#lx: Invalid vector element access in operand 0",
```
Comment (meme) [inline aarch64_extension.cpp:476]: Looks like a typo, did you mean `#%lx`?

File: aarch64_extension.cpp [hunk 4]
```diff
@@ -124,6 +210,42 @@ class AArch64ArchitectureExtension : public ArchitectureHook {
...
+    for (size_t Vd = 0; Vd < 32; Vd++) {
+      char baseRegisterName[3];
+
+      snprintf(baseRegisterName, sizeof(baseRegisterName), "v%zu", Vd);
+
+      for (auto& form : regForms) {
+        BNRegisterInfo regInfo;
+        char* regNameCStr = NULL;
+
+        if (asprintf(&regNameCStr, "v%zu%c[%zu]", Vd, form.elemType,
```
Comment (meme) [inline aarch64_extension.cpp:193]: This'll leak if I'm not mistaken

File: aarch64_extension.cpp [hunk 5]
```diff
@@ -124,6 +210,42 @@ class AArch64ArchitectureExtension : public ArchitectureHook {
...
+                     form.vectorIndex) < 0) {
+          LogError("asprintf: %s", strerror(errno));
+          /* XXX: should this be a C++ exception or something? */
+          assert(false);
```
Comment (meme) [inline aarch64_extension.cpp]: See note above

File: aarch64_extension.cpp [hunk 6]
```diff
@@ -97,6 +169,20 @@ class AArch64ArchitectureExtension : public ArchitectureHook {
...
+      if (reg.fullWidthRegister == baseRegister &&
+          reg.offset == elemSize * index && reg.size == elemSize) {
+        return id;
+      }
+    }
+    return 0;
+  }
+
 public:
   explicit AArch64ArchitectureExtension(Architecture* aarch64)
```
Comment (meme) [inline aarch64_extension.cpp:146]: I think this constructor has kind of blown up now, it might be a good idea to split this off into a `bool Init()` function and then handle all errors as a `return false;`, then when Init is called in the plug-in initializer, we can print an error and not load the plug-in. You are welcome to do this, if not I can make the adjustment after the PR is merged.

File: aarch64_extension.cpp [hunk 7]
```diff
@@ -45,11 +49,79 @@ class Disassembler {
 // called and the associated Capstone resources will be released
 static thread_local Disassembler disassembler;
 
+/*
+ * Given a Vector Arrangement Specifier, return the _element_ size to be
+ * operated on.
+ * Verify that the VAS is indeed valid before calling this function.
+ */
+static size_t vectorElementSize(enum arm64_vas vas) {
```
Comment (meme) [inline aarch64_extension.cpp]: Where is this used?

**Review by meme — CHANGES_REQUESTED** [review]

File: aarch64_extension.cpp (hunk 2 above)
Comment (nshp) [reply aarch64_extension.cpp]: We may want to `#define` that, but we're called for every instruction - we cannot return multiple instructions in one call, so it's a waste to have capstone disassemble the rest.

File: aarch64_extension.cpp (hunk 1 above)
Comment (nshp) [reply aarch64_extension.cpp:133]: Depends on how much we use it. I say don't make the data structures more complicated until it actually becomes a problem, unless you see an even cleaner way to do this.

File: aarch64_extension.cpp (hunk 2 above)
Comment (meme) [reply aarch64_extension.cpp]: `#define` sounds good.

File: aarch64_extension.cpp (hunk 3 above)
Comment (nshp) [reply aarch64_extension.cpp:476]: `%#lx` => `%`:format `#`:alternate `l`:long `x`:hex
The alternate modifier adds a `0x` prefix for us.

File: aarch64_extension.cpp (hunk 1 above)
Comment (meme) [reply aarch64_extension.cpp:133]: Seems reasonable, multi-conditionals makes it kind of annoying to use a hash-map here.

File: aarch64_extension.cpp (hunk 3 above)
Comment (meme) [reply aarch64_extension.cpp:476]: Very cool, I did not know that. TIL.

File: aarch64_extension.cpp (hunk 5 above)
Comment (meme) [reply aarch64_extension.cpp]: (I meant to reference https://github.com/meme/aarch64-extensions/pull/7#discussion_r414068234)

File: aarch64_extension.cpp (hunk 7 above)
Comment (nshp) [reply aarch64_extension.cpp]: Good catch, that was meant for the other branch. Removed.

File: aarch64_extension.cpp (hunk 6 above)
Comment (nshp) [reply aarch64_extension.cpp:146]: Agreed, done.

File: aarch64_extension.cpp (hunk 4 above)
Comment (nshp) [reply aarch64_extension.cpp:193]: Fixed.

**Comment by meme** [comment]
Lookin' good. Thanks for the PR. Probably about time we break this code up a bit though...
